### DISCUSSION TITLE
How do DataNodes handle messages when they share a same physical channel

### DISCUSSION THREAD

**Discussion by Reidddddd** [milvus-io/milvus#36369] [Q&A and General discussion] [answered]
Assuming there are two collections (a, b) , each one has 1 v channel, but they share a same p channel.

In datanodes side, each v channel will new a consumer, but consuming a same topic in this case. 

My question is: how does these datanodes handle messages interleaving? 
e.g  | - insert1 to a - | - insert1 to b - | - upsert2 to a - | ...*.  

Will any more difference when two consumers are on a same datanode?

**Comment by Reidddddd** [comment]
Please also correct me if my assumption is wrong already...

But I think the situation, sharing a same p channel, must exist.

Or is there another way to let consumer consume independently even sharing a same p channel? 

I couldn't find the answer via source codes (not straight forward at least), so here to ask directly for some hints and design details

Thanks

**Comment by yhmo** [comment]
Two collections [A, B], two v-channel, one p-channel. Each v-channel has a consumer.
Both messages of A and B enter the p-channel.
No matter the two consumers are in the same data node or different data nodes:
The consumer of A can read messages of B, it just ignores messages of B, and only consumes messages of A.
The consumer of B can read messages of A, it just ignores messages of A, and only consumes messages of B.

**Reply by Reidddddd** [reply]
Ok, checked, there is a `collection id` check.

Further question, what if a collection has two partitions, but share a same p-channel (set dml channel count 1).

Therefore the `collection id` check doesn't work, it this case, will two consumers consume a same message twice?

**Reply by yhmo** [reply, marked as answer]
The name of each v-channel is different. Each message knows its v-channel name. The proxy splits data of insert() into pieces according to shards numbers, and it knows which partition the data belongs to. Different partitions have different v-channels. Different v-channels have different names. The proxy sends the correct message with the correct v-channel name. In the data node, the consumer checks the message's v-channel name, if the name is equal, the message is consumed, otherwise ignored.

**Reply by Reidddddd** [reply]
Thx, I found the related source codes as you mentioned.
